File: ROS/catkin_hw/src/hello_world/src/listener.cpp

#include "ros/ros.h"
#include "std_msgs/String.h"


// 接收到订阅的消息，进入回调函数
void chatterCallback(const std_msgs::String::ConstPtr & msg)
{
	ROS_INFO("I heard: [%s]", msg->data.c_str()); // 将接收的消息打印出来
}

int main(int argc,char **argv)
{
	ros::init(argc,argv,"listener"); // 初始化一个名为listener的节点
	ros::NodeHandle n; // 创建名为n的节点句柄
	// 创建一个Subscriber的话题，注册回调函数chatterCallback
	ros::Subscriber sub=n.subscribe("chatter", 1000, chatterCallback);
	// 循环等待回调函数
	ros::spin();
	return 0;
}  

File: ROS/catkin_hw_py/src/helloworld/scripts/listener.py
#! /usr/bin/python
# -*- encoding: utf-8 -*-

import rospy
from std_msgs.msg import String

# Subscribe的回调函数
def callback(data):
    rospy.loginfo(rospy.get_caller_id() + 'I heard %s', data.data)

def listener():

    rospy.init_node('listener', anonymous=True)
    rospy.Subscriber('chatter', String, callback)
    rospy.spin()

if __name__ == '__main__':
    try:
        listener()
    except rospy.ROSInterruptException:
        pass